include/seqlock.h

```diff
@@ -0,0 +1,348 @@
+ * lock is for data where the reader wants a consistent set of information
+ * and is willing to retry if the information changes. There are two types
+ * of readers:
+ * 1. Sequence readers which never block a writer but they may have to retry
+ *    if a writer is in progress by detecting change in sequence number.
+ *    Writers do not wait for a sequence reader.
+ *
+ * This is not as cache friendly as brlock. Also, this may not work well
+ * for data that contains pointers, because any writer could
+ * invalidate a pointer that a reader was following.
+ *
+ * Expected non-blocking reader usage:
+ * 	do {
+ *	    seq = read_seqbegin(&foo);
+ * 	...
+ *      } while (read_seqretry(&foo, seq));
+ *
+ *
+ * On non-SMP the spin locks disappear but the writer still needs
+ * to increment the sequence variables because an interrupt routine could
+ * change the state of the data.
+ *
+ * Based on x86_64 vsyscall gettimeofday 
+ * by Keith Owens and Andrea Arcangeli
+ */
+
+static __always_inline void
+data_access_exceeds_word_size(void)
+{
+	RTE_LOG(WARNING, GATEKEEPER,
+		"seqlock: Data access exceeds word size and won't be atomic\n");
+}
+
+static __always_inline void
+__read_once_size(const volatile void *p, void *res, int size)
+{
+ 	switch (size) {
+ 	case 1:
+		*(uint8_t *)res = *(const volatile uint8_t*)p;
+		break;
+
+ 	case 2:
+		*(uint16_t *)res = *(const volatile uint16_t *)p;
+		break;
+
+	case 4:
+		*(uint32_t *)res = *(const volatile uint32_t *)p;
+		break;
+
+ 	case 8:
+		*(uint64_t *)res = *(const volatile uint64_t *)p;
+		break;
+
+ 	default:
+ 		data_access_exceeds_word_size();
+		break;
+	}
+}
+
+/*
+ * Prevent the compiler from merging or refetching reads or writes. The
+ * compiler is also forbidden from reordering successive instances of
+ * READ_ONCE, WRITE_ONCE and ACCESS_ONCE (see below), but only when the
+ * compiler is aware of some particular ordering.  One way to make the
+ * compiler aware of ordering is to put the two invocations of READ_ONCE,
+ * WRITE_ONCE or ACCESS_ONCE() in different C statements.
+ *
+ * In contrast to ACCESS_ONCE these two macros will also work on aggregate
+ * data types like structs or unions. If the size of the accessed data
+ * type exceeds the word size of the machine (e.g., 32 bits or 64 bits)
+ * READ_ONCE() and WRITE_ONCE()  will fall back to memcpy and print a
+ * compile-time warning.
+ *
+ * Their two major use cases are: (1) Mediating communication between
+ * process-level code and irq/NMI handlers, all running on the same CPU,
+ * and (2) Ensuring that the compiler does not  fold, spindle, or otherwise
+ * mutilate accesses that either do not require ordering or that interact
+ * with an explicit memory barrier or atomic instruction that provides the
+ * required ordering.
+ */
+#define READ_ONCE(x) \
+	({ union { typeof(x) __val; char __c[1]; } __u; \
+	__read_once_size(&(x), __u.__c, sizeof(x)); __u.__val; })
+
+#define cpu_relax() asm ("rep; nop" ::: "memory")
```

Comment:
You drop `cpu_relax()` in favor of `rte_pause()`, which we've already used in Gatekeeper.